### PR TITLE
Lf get all puzzles filter by name

### PR DESCRIPTION
**Pull request by luigiferiani**
Adding the option to filter by puzzle name when getting all puzzles from the database. 
This is a possible way to address #68

Also putting api/common code related tests in its own `tests/api` folder so that individual puzzle tests can live in `tests/puzzles`, mimicking the app's folder structure
```
openday_scavenger/
L api/
  L visitors/
  L puzzles/
L puzzles/
  L demo/
  L ...
```
with
```
tests/
L api/
  L visitors_tests/
  L puzzles_tests/
L puzzles/
  L demo_tests/
  L ...
```

### CODE REVIEW

**Review by portrain — COMMENTED** [review]
Very nice addition of a name filter. May I suggest to change the parameter name to reflect that it filters for puzzle names that start with the string (maybe something like `puzzle_name_startswith` or `filter_by_name_startswith`)

**Comment by luigiferiani** [comment]
I was actually **very** tempted to make the filtering an exact match, the only reason it's a "starts with" filter is because another service call was doing the same thing. But this is probably more flexible. Change done :)

**Comment by portrain** [comment]
It would probably be very useful to have a function with a signature similar to `get(db_session: Session, *, name: str) -> Puzzle` which returns a single puzzle by exact name match (using the fact here that `name` is the unique discriminator between puzzles)

**Comment by luigiferiani** [comment]
On it. Happy if I use this same branch/PR? not the cleanest but I guess I can change the PR title once I implement it. 
~Should it raise if there's no match, or just return None?~ It will raise, for consistency with the current behaviour of literally every other service function that does the equivalent of `get(db_session: Session, *, name: str) -> Puzzle` in its body...

**Comment by portrain** [comment]
Absolutely beautiful

**Comment by luigiferiani** [comment]
Hi @portrain (or @stmudie!), does this PR need any more work?

File: openday_scavenger/api/puzzles/service.py [hunk 1]
```diff
@@ -38,13 +38,21 @@
...
+) -> list[Puzzle]:
     """
-    Return all puzzles in the database, optionally only the active ones.
+    Return all puzzles in the database, with optional filtering.
+    optionally only the active ones.
 
     Args:
         db_session (Session): The SQLAlchemy session object.
         only_active (bool): Set this to True to only return active puzzles.
+        filter_by_puzzle_name (str): Only return responses
```
Comment (stmudie) [inline openday_scavenger/api/puzzles/service.py]: Docstring needs to be updated with change in argument name.

**Comment by stmudie** [comment]
Looks good to me. Just left very minor comment for a fix.

**Review by stmudie — APPROVED** [review]
Excellent work!!

**Pull request merged**